### PR TITLE
feature: Connect automatically when `.bsp` folder shows up

### PR DESCRIPTION
**Pull request by tgodzik**
Previously, if a user opened metals and then run .bsp generation they would need to restart metals or manually connect to build server. Now, we start watching for .bsp when no build tool exists and when it shows up we connect to the build server defined there.

### CODE REVIEW

File: metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala [hunk 1]
```diff
@@ -221,7 +221,7 @@ class MetalsLanguageServer(
       params => didChangeWatchedFiles(params),
     )
   )
-  private val indexingPromise: Promise[Unit] = Promise[Unit]()
+  val indexingPromise: Promise[Unit] = Promise[Unit]()
```
Comment (dos65) [inline metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala:224]: is this change needed?

File: metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala [hunk 2]
```diff
@@ -1426,6 +1426,12 @@ class MetalsLanguageServer(
   ): CompletableFuture[Unit] = {
     val path = AbsolutePath(event.path)
     val isScalaOrJava = path.isScalaOrJava
+
+    event.eventType match {
+      case EventType.CreateOrModify if path.isBsp =>
+        quickConnectToBuildServer()
```
Comment (dos65) [inline metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala:1432]: I was thinking that probably it would be nice if we start generating `.bsp` for scala-cli from Metals as an option for importing build if there no other build tools were detected. Also, the same approach might be applied for sbt and mill. For example if we see that it's sbt we can suggest two options for import even if there is no `.bsp` directory.

Maybe it would be better to improve the importing logic rather than adding this handler? wdyt?
I think it will lead to duplicated reconnect calls.

File: metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala (hunk 1 above)
Comment (tgodzik) [reply metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala:224]: That's needed for the test unfortunately

File: metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala (hunk 2 above)
Comment (tgodzik) [reply metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala:1432]: It doesn't lead to duplicated reconnection from what I checked. We only listen for .BSP only if we don't find any build tool.

I don't think we should yet propose scalacli as default(not sure if it wouldn't be flaky) and this isn't actually exclusive. People might run scalacli manually and it's sensible to connect to it in that case.

Comment (tgodzik) [reply metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala:1432]: Improving importing logic might also be good, but it's not necessarily related to this and doesn't solve the issue of what happens if BSP was generated outside of metals. People shouldn't need to restart metals in that case.

**Comment by tgodzik** [comment]
> Does that mean, Mill users can immediately start using `mill-bsp`?. Until now, it is necessary to first start a bloop-based setup, just to switch afterwards to `mill-bsp`.

If you generate .BSP directory metals will always try to connect to it first. This however doesn't change the behaviour when no .bsp is present. We might also want to improve that but it's not strictly related

**Review by alexarchambault — APPROVED** [review]
AFAIC, looks good!

**Comment by kpodsiad** [comment]
> > Does that mean, Mill users can immediately start using `mill-bsp`?. Until now, it is necessary to first start a bloop-based setup, just to switch afterwards to `mill-bsp`.
> 
> If you generate .BSP directory metals will always try to connect to it first. This however doesn't change the behaviour when no .bsp is present. We might also want to improve that but it's not strictly related

@tgodzik quite recently I've tried mill-bsp in a fresh project and even though I had mill entry under `.bsp` Metals weren't able to autoconnect to it. By default Metals were trying to connect to the bloop which was ending in `build server is not autoconnectable`. Only after `switch build server` and explicit choice of `mill` `connect to the build server` worked.

**Comment by lefou** [comment]
> > > Does that mean, Mill users can immediately start using `mill-bsp`?. Until now, it is necessary to first start a bloop-based setup, just to switch afterwards to `mill-bsp`.
> > 
> > 
> > If you generate .BSP directory metals will always try to connect to it first. This however doesn't change the behaviour when no .bsp is present. We might also want to improve that but it's not strictly related
> 
> @tgodzik quite recently I've tried mill-bsp in a fresh project and even though I had mill entry under `.bsp` Metals weren't able to autoconnect to it. By default Metals were trying to connect to the bloop which was ending in `build server is not autoconnectable`. Only after `switch build server` and explicit choice of `mill` `connect to the build server` worked.

That's exactly what I experienced. Always. Hence my initial questions.

File: metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala (hunk 2 above)
Comment (tgodzik) [reply metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala]: Also added a check for bspSession to make sure it's empty.

**Comment by tgodzik** [comment]
> > > > Does that mean, Mill users can immediately start using `mill-bsp`?. Until now, it is necessary to first start a bloop-based setup, just to switch afterwards to `mill-bsp`.
> > > 
> > > 
> > > If you generate .BSP directory metals will always try to connect to it first. This however doesn't change the behaviour when no .bsp is present. We might also want to improve that but it's not strictly related
> > 
> > 
> > @tgodzik quite recently I've tried mill-bsp in a fresh project and even though I had mill entry under `.bsp` Metals weren't able to autoconnect to it. By default Metals were trying to connect to the bloop which was ending in `build server is not autoconnectable`. Only after `switch build server` and explicit choice of `mill` `connect to the build server` worked.
> 
> That's exactly what I experienced. Always. Hence my initial question.

Och, that looks like a bug then :/

File: metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala (hunk 2 above)
Comment (tgodzik) [reply metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala:1432]: Ach actually that caused issues with tests. I think it's fine to reconnect whenever `.bsp` changes since it need to be changed explicitely.

**Review by dos65 — APPROVED** [review]

**Comment by ckipp01** [comment]
> Och, that looks like a bug then :/

Actually originally this was the desired behavior. Now that we have more viable servers we may want to change this. I believe we do a check and if a build tool is supported by Bloop no matter what we defaulted to this. We can look into changing this behavior, but that would also mean that for a lot of sbt users, if they run sbt before they ever open Metals, then. they'd be prompted to auto-connect to sbt instead of Bloop. Is it time that we go that direction?

**Comment by lefou** [comment]
> > Och, that looks like a bug then :/
> 
> Actually originally this was the desired behavior. Now that we have more viable servers we may want to change this. I believe we do a check and if a build tool is supported by Bloop no matter what we defaulted to this. We can look into changing this behavior, but that would also mean that for a lot of sbt users, if they run sbt before they ever open Metals, then. they'd be prompted to auto-connect to sbt instead of Bloop. Is it time that we go that direction?

Is sbt creating a `.bsp/` dir unasked? At least for Mill, we would only trigger such behavior once the `mill.bsp.BSP/install` command was invoked (explicitly by the user).

**Comment by tgodzik** [comment]
Yeah, it always creates .bsp unasked.

I think the best current solution is to suggest both bloop and sbt/mill server at the same time of .bsp is present and the user hasn't chosen .

What do you think?

**Comment by ckipp01** [comment]
>  What do you think?

I think there are some pros and cons to this. If I understand you correctly in a scenario where there is a `.bsp/build-tool.json` and that build tool supports bloop, then upon the first opening of the project the user would asked to use `build-tool` or Bloop as their build server.

The pro to this is that in the situation @lefou outlined, you could use Mill right away. The con is that for a new user, this might confuse them if they get an option to use sbt or Bloop for example. I'm not really sure.

**Comment by lefou** [comment]
IMHO, the right behavior would be to just use what's defined in `.bsp/build-tool.json`. There is a reason this dir/file exists. We should only ask, if there is either no tool json at all or  it found more that one tool json in this dir. Problem with this approach is sbt, which is not behaving nicely here. Question is, whether this is on purpose or only because they can? Maybe, we can convince sbt folks to change and only create the file on demand? I have exactly no internal insights, so this is a rather personal opinion. This is definitely a topic, where BSP should be the leader and make suggestions how servers and clients should behave.

**Comment by ckipp01** [comment]
> This is definitely a topic, where BSP should be the leader and make suggestions how servers and clients should behave.

I agree.

**Comment by dos65** [comment]
I don't think that there is any issue with any approach - create by default or on demand. The current Metals behavior is anyway inconsistent. Let's say you check out a repo and open it in editor without interacting with build tool. In this case there will be no difference between sbt and mill - the only suggested option will be bloop.
 
I believe even if there is no `.bsp` it would be better to provide both option and generate `.bsp` from Metals side if it's necessary.
